util/loadCollections.ts

```diff
@@ -2,59 +2,80 @@ import { promises as fs } from 'fs'
 
 import yaml from 'js-yaml'
 
+import AuthorCollectionDefinition from './AuthorCollectionDefinition'
 import loadAllRecords from './loadAllRecords'
 import MarkdownFile from './MarkdownFile'
+import RecordCollectionDefinition from './RecordCollectionDefinition'
+import { normalizeRecordCollectionDefinition, normalizeAuthorCollectionDefinition } from './validators'
 
-interface CollectionDefinition {
+export interface RecordCollection {
     title: string
-    slug: string
+    slug?: string
     type: string
-    members: string[]
+    members: MarkdownFile[]
 }
 
-export interface AuthorDefinition {
-    id: string
-    name: string
-    bio: string
-    socialLinks: string[]
+interface Collections {
+    recordCollections: RecordCollection[]
+    authors: AuthorCollectionDefinition[]
 }
 
-export interface RecordCollection {
-    title: string
-    slug: string
-    members: MarkdownFile[]
+function findMemberRecord(records: MarkdownFile[], memberSlug: string, collectionDefinition: RecordCollectionDefinition): MarkdownFile {
+    const memberRecord = records.find(record => record.slug === memberSlug)
+    if (!memberRecord) {
+        throw new Error(
+            `Record not found: "${memberSlug}" from collection "${collectionDefinition.title})"`
+        )
+    }
+    return memberRecord
 }
 
-interface Collections {
-    recordCollections: RecordCollection[]
-    authors: AuthorDefinition[]
+function returnAllMemberRecords(collectionDefinition: RecordCollectionDefinition, records: MarkdownFile[]):MarkdownFile[] {
+    let uniqueMemberRecords: MarkdownFile[] = []
+    collectionDefinition.members.map(memberSlug => {
+        const memberRecord = findMemberRecord(records, memberSlug, collectionDefinition)
+        uniqueMemberRecords = [ ...uniqueMemberRecords, memberRecord ]
+    })
+
+    return uniqueMemberRecords
+}
+
+function returnRecordCollections(collections: RecordCollectionDefinition[], records: MarkdownFile[]): RecordCollection[] {
```

Comment:
Refactoring things this way made sense to me for two reasons: 1. It enabled the yaml validation in a place that I thought made sense (before we fetch records); and 2. I thought it would be easier down the road to embed a record type check here.

Breaking these loops into more methods also means more code (le sigh), but hopefully there is a readability boost?

Reply:
+1 readability